File: atlas/examples/kotlin/app/src/main/kotlin/dev/atlas/example/app/CatalogApplication.kt
package dev.atlas.example.app

import dev.atlas.example.application.catalog.CatalogQueryService
import dev.atlas.example.application.catalog.ImportCatalog
import dev.atlas.example.domain.catalog.CatalogItemFactory
import dev.atlas.example.domain.catalog.DefaultCatalogTitlePolicy
import dev.atlas.example.infrastructure.catalog.DEFAULT_CATALOG_SEED
import dev.atlas.example.infrastructure.catalog.FixedCatalogClock
import dev.atlas.example.infrastructure.catalog.InMemoryCatalogRepository
import dev.atlas.example.infrastructure.catalog.JsonCatalogSeedReader
import java.time.Instant

/**
 * Owns concrete dependency composition for the runnable catalog application.
 */
class CatalogApplication private constructor(
    private val demo: CatalogDemo
) {
    /**
     * Runs the composed catalog workflow.
     */
    fun run() {
        demo.run()
    }

    companion object {
        /**
         * Composes the deterministic example adapters and use cases.
         *
         * @return Ready-to-run catalog application.
         */
        fun createDefault(): CatalogApplication {
            val repository = InMemoryCatalogRepository()
            val itemFactory = CatalogItemFactory(DefaultCatalogTitlePolicy())
            val seedReader = JsonCatalogSeedReader(DEFAULT_CATALOG_SEED)
            val clock = FixedCatalogClock(Instant.parse("2026-07-30T00:00:00.000Z"))
            val importCatalog = ImportCatalog(itemFactory, repository, seedReader, clock)
            val queryService = CatalogQueryService(repository)
            val service = ApplicationCatalogService(importCatalog, queryService)
            return CatalogApplication(CatalogDemo(service, RuntimeOutputWriter()))
        }
    }
}

File: atlas/examples/typescript/packages/app/src/composition/CatalogApplication.ts
import { CatalogQueryService, ImportCatalog } from '@atlas-example/application';
import {
  CatalogItemFactory,
  DefaultCatalogTitlePolicy,
} from '@atlas-example/domain';
import {
  DEFAULT_CATALOG_SEED,
  FixedCatalogClock,
  InMemoryCatalogRepository,
  JsonCatalogSeedReader,
} from '@atlas-example/infrastructure';
import { CatalogDemo } from '../CatalogDemo.js';
import { RuntimeOutputWriter } from '../RuntimeOutputWriter.js';

/**
 * Owns concrete dependency composition for the runnable catalog application.
 */
export class CatalogApplication {
  /**
   * Creates the executable application around its delivery workflow.
   *
   * @param demo - Fully composed catalog presentation workflow.
   */
  private constructor(private readonly demo: CatalogDemo) {}

  /**
   * Composes the deterministic example adapters and use cases.
   *
   * @returns Ready-to-run catalog application.
   */
  public static createDefault(): CatalogApplication {
    const repository = new InMemoryCatalogRepository();
    const itemFactory = new CatalogItemFactory(new DefaultCatalogTitlePolicy());
    const seedReader = new JsonCatalogSeedReader(DEFAULT_CATALOG_SEED);
    const clock = new FixedCatalogClock(new Date('2026-07-30T12:00:00.000Z'));
    const importCatalog = new ImportCatalog(
      itemFactory,
      repository,
      seedReader,
      clock,
    );
    const queryService = new CatalogQueryService(repository);
    return new CatalogApplication(
      new CatalogDemo(importCatalog, queryService, new RuntimeOutputWriter()),
    );
  }

  /**
   * Runs the composed catalog workflow.
   */
  public run(): void {
    this.demo.run();
  }
}
